### PR TITLE
feat: Add KV events for Mamba radix cache

### PR DESCRIPTION
**Pull request by zhongdaor-nv**
## Motivation

`MambaRadixCache` currently inherits `BasePrefixCache.take_events()`, so enabling `--kv-events-config` does not publish KV cache placement events for the Mamba radix path. Downstream KV-aware routers then lose precise cache placement information and fall back to approximate routing.

## Changes

- Initialize the KV event queue in `MambaRadixCache` and implement `take_events()`.
- Emit `BlockStored` for inserted full-attention KV nodes.
- Emit `BlockRemoved` when full-attention KV blocks are actually evicted, including tombstone cleanup.
- Keep Mamba-only tombstoning silent because the existing KV event schema represents full-attention KV blocks, not Mamba state groups.
- Preserve page hash values across node splits so parent/child event hashes remain stable.
- Add unit coverage for store, full eviction, Mamba-only tombstoning, and split parent hash behavior.

## Tests

- `source ~/sgl/bin/activate && PYTHONPATH=python python -m pytest test/registered/unit/mem_cache/test_mamba_unittest.py -q`
- `source ~/sgl/bin/activate && PYTHONPATH=python python -m compileall -q python/sglang/srt/mem_cache/mamba_radix_cache.py test/registered/unit/mem_cache/test_mamba_unittest.py`
- `git diff --check`

## Validation

End-to-end test with **NVIDIA Dynamo + SGLang on Qwen3.5-4B**, single
GPU (RTX 6000 Ada), 2 sglang workers sharing the GPU, kv-events enabled:

| Check | Observed |
|---|---|
| `BlockStored` events emitted under prefill | ✅  |
| `BlockRemoved` events emitted under cache pressure | ✅  19,196 events observed in a 320-request cache-pressure workload |
| `AllBlocksCleared` emitted on `reset()` | ✅ |
| Events forwarded over ZMQ → dynamo KV indexer | ✅  decoded with msgspec, schema matches |
| Dynamo router uses indexer state for routing | ✅  in repeat-prompt scenarios, SGLang's `usage.prompt_tokens_details.cached_tokens > 0`, confirming the request landed on the worker that previously cached the prompt |

Able to observed Kv events 

```
BlockStored block_hashes=[7417413287883245477] parent=4297089568659962000 block_size=16 medium=GPU n_tokens=16
BlockStored block_hashes=[9181454537587020687] parent=7417413287883245477 block_size=16 medium=GPU n_tokens=16
BlockStored block_hashes=[7262263677522409942] parent=9181454537587020687 block_size=16 medium=GPU n_tokens=16
....
[tcp://127.0.0.1:5558] seq=3 ts=1777355568.544165 dp=0 n_events=744
BlockRemoved block_hashes=[2483298193547563680] medium=GPU
BlockRemoved block_hashes=[-5185890641737854172] medium=GPU
```

### CODE REVIEW

**Comment by zhongdaor-nv** [comment]
### Dynamo kv aware routing demo: same 5 prompts sent twice  (Qwen3.5-4B)
  
  `Qwen3.5-4B` on 1× RTX 6000 Ada (2 sglang workers sharing GPU, mem-fraction 0.42 each). 5 distinct ~3300-token prompts × 2 rounds, `max_tokens=1` so latency reflects prefill cost.
  
  **Demonstration only and not a benchmark.** The numbers below are from a single run on a minimal setup (1 GPU, 2 sglang workers sharing it, 5 synthetic prompts, sequential single-frontend) — they are meant only to show that the events emitted by this PR enable dynamo's KV-aware path to work end-to-end on a hybrid SSM model.
    | prefix | mode | round 1 → worker | r1 lat | r1 cached | round 2 → worker | r2 lat | **r2 cached** |
  |---|---|---|---|---|---|---|---|
  | A | **KV-aware (PR)** | worker 1 | 310 ms | — | worker 1 | **137 ms** | ✓ **3136 / 3188** |
  |   | round-robin | worker 1 | 328 ms | — | **worker 2** | 317 ms | ✗ — |
  | B | **KV-aware (PR)** | worker 2 | 296 ms | — | worker 2 | **111 ms** | ✓ **3136 / 3188** |
  |   | round-robin | worker 2 | 322 ms | — | **worker 1** | 316 ms | ✗ — |
  | C | **KV-aware (PR)** | worker 1 | 327 ms | — | worker 1 | **136 ms** | ✓ **3136 / 3188** |
  |   | round-robin | worker 1 | 301 ms | — | **worker 2** | 286 ms | ✗ — |
  | D | **KV-aware (PR)** | worker 2 | 299 ms | — | worker 2 | **133 ms** | ✓ **3136 / 3188** |
  |   | round-robin | worker 2 | 302 ms | — | **worker 1** | 294 ms | ✗ — |
  | E | **KV-aware (PR)** | worker 2 | 322 ms | — | worker 2 | **111 ms** | ✓ **3136 / 3188** |
  |   | round-robin | worker 1 | 308 ms | — | **worker 2** | 316 ms | ✗ — |

  | | round 1 (cold) | round 2 (repeat) | speedup | r2 cache reused |
  |---|---|---|---|---|
  | **KV-aware (this PR)** | 311 ms | **126 ms** | **2.47×** | **98.4 %**  (3136 / 3188 tokens × 5/5 hits) |
  | round-robin (default) | 312 ms | 306 ms | 1.02× | 0 %  (0 / 3188 tokens × 0/5 hits) |

  **Mechanism:** with the events emitted by this PR, dynamo's KV indexer knows which worker holds each prompt's KV cache, so the router sends each repeat request back to that worker (`w_9618` → `w_9618`). Without this kv aware routing, dynamo use `round-robin` by default, which strictly alternates workers (`w_9670` → `w_9668`) and forces each repeat to re-prefill 3300 tokens on a worker that hadn't seen the prompt before.

File: python/sglang/srt/mem_cache/mamba_radix_cache.py [hunk 1]
```diff
@@ -1220,6 +1241,87 @@ def _delete_tombstone_leaf(self, node: TreeNode) -> None:
 
         self.full_evictable_size_ -= len(node.key)
 
+    def _record_store_event(self, node: TreeNode, medium=None):
```
Comment (hzh0425) [inline python/sglang/srt/mem_cache/mamba_radix_cache.py]: Seems like this _record_xxx_event methods are almost the same with RadixTree

Could we abstract this method into utils for direct reuse?

File: test/registered/unit/mem_cache/test_mamba_unittest.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import torch
```
Comment (hzh0425) [inline test/registered/unit/mem_cache/test_mamba_unittest.py:3]: Please also add a server test in test/registered/4-gpu-models/test_qwen35_hicache.py

File: python/sglang/srt/mem_cache/mamba_radix_cache.py [hunk 3]
```diff
@@ -1088,6 +1104,9 @@ def _split_node(self, key: RadixKey, child: TreeNode, split_len: int) -> TreeNod
         child.key = child.key[split_len:]
         child.value = child.value[split_len:].clone()
         new_node.parent.children[key.child_key(self.page_size)] = new_node
+        new_node.hash_value, child.hash_value = split_node_hash_value(
```
Comment (hzh0425) [inline python/sglang/srt/mem_cache/mamba_radix_cache.py:1099]: This conflicts with HiMambaRadixTree._split_node.

Additionally, Mamba HiCache is currently supported within HiMambaRadixTree. Could you please help verify whether the newly added KV event feature is compatible with HiMambaRadixTree?

File: python/sglang/srt/mem_cache/mamba_radix_cache.py (hunk 1 above)
Comment (ishandhanani) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py]: +1

Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py]: done. now they are all in utils.py

File: test/registered/unit/mem_cache/test_mamba_unittest.py (hunk 2 above)
Comment (zhongdaor-nv) [reply test/registered/unit/mem_cache/test_mamba_unittest.py:3]: done

File: python/sglang/srt/mem_cache/mamba_radix_cache.py (hunk 3 above)
Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py:1099]: Good catch. Addressed in 246f0b4d. HiMambaRadixCache._split_node no longer carries its own split_node_hash_value call; it now delegates to super()._split_node()
  
  End-to-end coverage on the hicache path. The new server test test/registered/4-gpu-models/test_qwen35_hicache.py runs Qwen3.5-27B with --enable-hierarchical-cache
  --kv-events-config '{"publisher":"zmq","topic":"kv-events"}', which routes through HiMambaRadixCache

**Comment by zhongdaor-nv** [comment]
Hi @hzh0425, I’ve addressed the comments. Please take another look when you get a chance.

File: python/sglang/srt/mem_cache/utils.py [hunk 4]
```diff
@@ -389,3 +395,146 @@ def hash_str_to_int64(hash_str: str) -> int:
...
+        # One BlockStored per ``page_size`` chunk.
+        # ``medium`` defaults to StorageMedium.GPU but callers may override
+        # for lower-tier insertions (e.g. StorageMedium.CPU for host/L2 cache).
+        if self.enable_kv_cache_events:
+            if medium is None:
+                medium = StorageMedium.GPU
+
+            # Compute hash_value lazily if not already set
+            if node.hash_value is None:
+                node.hash_value = compute_node_hash_values(node, self.page_size)
```
Comment (hzh0425) [inline python/sglang/srt/mem_cache/utils.py]: hi @zhongdaor-nv 

I feel like we probably can't modify the tree's properties from outside the tree.

Does KVEEvent only take effect when enable_storage is enabled? And does the Tree itself guarantee that the node's hash value has already been computed when enable_storage?

File: python/sglang/srt/mem_cache/mamba_radix_cache.py (hunk 3 above)
Comment (hzh0425) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py:1099]: the question is, Why do we need to split the hash value here instead of inside HiMambaTree? It doesn't seem like there are any dependencies requiring this.

Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py:1099]: Before this PR, MambaRadixCache didn't track node.hash_value at all. This PR introduces node.hash_value on MambaRadixCache so it can publish KV events. 

Once hash_value exists on the node, it's just another prefix-aligned per-token array alongside node.key and node.value, so _split_node has to slice it at the same split_len for the same reason it already slices key and value.

Moving this into HiMambaRadixCache only would mean MambaRadixCache carries a hash_value field that _split_node silently leaves stale , which would break the plain (non-hicache) --kv-events-config path that this PR is built around.

Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/mamba_radix_cache.py:1099]: More speficically, this PR is scoped to emitting KV events when --kv-events-config is set, and is not intended to change anything related to HiCache.

File: python/sglang/srt/mem_cache/utils.py (hunk 4 above)
Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/utils.py]: Strictly speaking the mixin is part of the class via MRO, so it's not technically "outside"
There's similar pattern in sglang with class `Scheduler(SchedulerOutputProcessorMixin, SchedulerProfilerMixin, ..., SchedulerMetricsMixin...) 
`that mutation scheduler state.

Are you suggesting KVCacheEventMixin should be a pure emitter and not mutate node state? If so, do you prefer it to assert/raise when node.hash_value is None, or to silently early return? Happy to align either way.

Comment (hzh0425) [reply python/sglang/srt/mem_cache/utils.py]: Everything else mostly makes sense.
The main concern is here: can we ensure that the tree's attributes are only modified within the tree itself, and that this location only captures events?

For example, when kv_event_config is set, we could control the generation and handling of node_hash_value right at the place in the tree where enable_storage is enabled.

Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/utils.py]: this lazy-compute pattern already existed on main inside RadixCache._record_store_event / _record_remove_event itself (radix_cache.py:866-868 / 915-917). The PR just relocated the same behavior into the mixin when extracting it for reuse. Want to minimize the change to reduce the potential of breaking things.

Let's sync offline on that.

Comment (hzh0425) [reply python/sglang/srt/mem_cache/utils.py]: ic

**Comment by hzh0425** [comment]
Let's move on the ci

**Review by hzh0425 — APPROVED** [review]
Looks good to me overall; could you folks double-check it?
@ishandhanani @xiezhq-hermann

File: python/sglang/srt/mem_cache/utils.py [hunk 5]
```diff
@@ -389,3 +395,146 @@ def hash_str_to_int64(hash_str: str) -> int:
...
+    else:
+        split_pages = split_len // page_size
+
+    new_node_hash = child_hash_value[:split_pages]
+    child_hash = child_hash_value[split_pages:]
+
+    return new_node_hash, child_hash
+
+
+class KVCacheEventMixin:
```
Comment (ishandhanani) [inline python/sglang/srt/mem_cache/utils.py]: nit: can we move this to `mem_cache/events.py`. Don't know if I like this mixin living in a utils file

File: test/registered/unit/mem_cache/test_mamba_unittest.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
...
 
 from sglang.srt.configs.mamba_utils import Mamba2CacheParams, Mamba2StateShape
+from sglang.srt.disaggregation.kv_events import BlockRemoved, BlockStored
```
Comment (ishandhanani) [inline test/registered/unit/mem_cache/test_mamba_unittest.py:6]: not blocker for this pr. 

note for a follow up pr: we should rlly move kv events to `mem_cache/`. I don't know why it is in `disaggregation` lol

**Review by ishandhanani — APPROVED** [review]

File: python/sglang/srt/mem_cache/utils.py (hunk 5 above)
Comment (zhongdaor-nv) [reply python/sglang/srt/mem_cache/utils.py]: done

**Comment by zhongdaor-nv** [comment]
Verified again e2e with a local Qwen3.5-0.8B sglang server and a ZMQ subscriber. After /generate (×21) and /flush_cache:

- 29,908 BlockStored (medium=GPU, parent_block_hash chain present)
- 28,500 BlockRemoved (medium=GPU)
- 1 AllBlocksCleared
- 28,399 of the removed hashes match the stored set, confirming store/remove closure.

**Comment by zhongdaor-nv** [comment]
Local unit tests: 72 passed, 0 failed (source ~/sgl/bin/activate && PYTHONPATH=python python -m pytest test/registered/unit/mem_cache/ -q)

Directly relevant to this PR:
- test_mamba_unittest.py::TestMamba::test_mamba_radix_cache_kv_events — store + evict event flow on MambaRadixCache
- test_mamba_unittest.py::TestMamba::test_mamba_radix_cache_kv_events_split_hash — pins parent_block_hash chain stability across _split_node
- test_radix_cache_unit.py::TestRadixCache::test_kv_cache_events / test_kv_cache_events_with_eviction — RadixCache event emission via the same KVCacheEventMixin
- test_radix_cache_unit.py::TestRadixCache::test_hash_value_split / test_hash_value_storage / test_hash_value_repeating_tokens — hash-chain invariants

Plus 64 sibling tests in test_mamba_unittest.py, test_radix_cache_unit.py, test_swa_unittest.py, test_evict_policy.py, test_radix_cache_slru_accuracy.py to confirm no regression in the surrounding cache code.

**Comment by hzh0425** [comment]
let's move on the ci
